Added basic TS definitions

JasonStoltz opened the pull request:
This add TS definition files to this library.

Closes #13.

This is published with version `@elastic/app-search-node@7.16.1-beta.1` for testing.

```
npm i --save @elastic/app-search-node@7.16.1-beta.1
```

This is the code I used to test these definitions: https://gist.github.com/JasonStoltz/9349e76a4d54607591a3258cfdabc7b8.

These aren't the most comprehensive types in terms of covering all API request and response bodies, users will still have to rely on documentation to some degree. This is primarily since these definitions were hand coded and NOT generated from our Open API spec.

The long term goal is to replace this client with a more comprehensive Enterprise Search client, which is generated from our Open API specs, but for the short term, this will keep this client useable.

Thank you @richbai90 for providing the basis for these types: https://github.com/elastic/app-search-node/issues/13#issuecomment-849176202

![latest](https://user-images.githubusercontent.com/1427475/145884334-ea94d9af-80be-475c-94b5-b86e1a6b2c78.gif)

JasonStoltz committed:
Basic type defs

JasonStoltz committed:
Added multi_search

JasonStoltz committed:
More defs

JasonStoltz committed:
All defs

JasonStoltz committed:
Remove inline jsdoc

JasonStoltz committed:
Updated examples

JasonStoltz committed:
Add client

JasonStoltz committed:
7.15.1-beta.0

JasonStoltz committed:
7.15.1-beta.1

JasonStoltz committed:
7.15.2-beta.0

JasonStoltz committed:
Merge branch 'master' into ts-definitions

byronhulcher commented on README.md:
```diff
@@ -109,341 +128,402 @@ Note that this API will not throw on an indexing error. Errors are inlined in th
-  .then(response => console.log(response))
-  .catch(error => console.log(error))
+} catch (e) {
+  console.log(e);
```
```suggestion
  console.error(e);
```

byronhulcher commented on README.md:
```diff
@@ -109,341 +128,402 @@ Note that this API will not throw on an indexing error. Errors are inlined in th
+  );
+  console.log(`Indexed ${singleDocumentResponse.id} succesfully`);
+} catch (e) {
+  console.log(e);
```
```suggestion
  console.error(e);
```

byronhulcher commented on app-search-node.d.ts:
```diff
@@ -0,0 +1,725 @@
+     * const client = new AppSearchClient(hostIdentifier, apiKey)
+     */
+    constructor(
+      accountHostKey: string | undefined,
```
Can this be `accountHostKey?: string`? Not sure if the compiler will complain about optional arguments preceding required arguments but wanted to raise.

JasonStoltz replied:
The compiler *does* complain about the optional arguments preceding required ones. Maybe with the upcoming 8.0 release we can change this signature to be better.

byronhulcher commented on app-search-node.d.ts:
```diff
@@ -0,0 +1,725 @@
+     *     );
+     *   }
+     * } catch (e) {
+     *   console.log(e);
```
```suggestion
     *   console.error(e);
```

byronhulcher commented on lib/appSearch.js:
```diff
@@ -43,27 +43,11 @@ class AppSearchClient {
     this.client = new Client(apiKey, baseUrl)
   }
 
-  /**
-   * Send a search request to the App Search Api
-   * https://swiftype.com/documentation/app-search/api/overview
-   *
-   * @param {String} engineName unique Engine name
-   * @param {String} query String that is used to perform a search request.
-   * @param {Object} options Object used for configuring the search like search_fields and result_fields
-   * @returns {Promise<Object>} a Promise that returns a result {Object} when resolved, otherwise throws an Error.
-   */
```
Its kind of a bummer to lose these helpful code comments

JasonStoltz replied:
Which ones, the parameter comments? I could add them to the definition files.

JasonStoltz replied:
And are you thinking about comments for a consumer of this library, or for developers?

byronhulcher replied:
Yeah, maybe with some discretion? I'm sure a developer can probably guess what `engineName: string` is used for haha.  On the other hand, maybe types make this sort of thing "self documenting"? I'd handle in a follow-up either way, that work shouldn't block this PR in my opinion

byronhulcher replied:
> And are you thinking about comments for a consumer of this library, or for developers?

A customer/developer integrating `app-search-node` into their app.  

Looking into this a little, I think you should be able to copy/paste them into our typecript files, I get the impression `tsdoc` and `jsdoc` share syntax.  See https://tsdoc.org/

JasonStoltz replied:
I think you're right. And that makes sense, I'll do this in a follow-up.

JasonStoltz replied:
@byronhulcher I just went ahead and re-added it: https://github.com/elastic/app-search-node/pull/50/commits/32b894e56d2eb2c441801384f31a55b06fe1fa75

byronhulcher reviewed (dismissed):
Looks good! I flagge a couple of places I think you might have wanted to use `console.error` instead of `console.log` in `catch` blocks, that's the only issues I could find.

One comment is that there's a lot of places you use `Record<string, any>` to represent a type you're gonna fill in later (documents, results, options, etc).  This makes it a little hard to work out from looking at the types what these represent.  If I was going about this, I'd probably create something like `type DocumentResponse = Record<string, any>` and `type MultiSearchOptions = Record<string, any>`.  That may be overkill, especially if you're going to do a follow-up fleshing those out more.  Maybe something to consider for the next module you add types to.

JasonStoltz committed:
Fix console.log statements

JasonStoltz committed:
A few fixes:

- Added tsdoc param descriptions back
- Added missing querySuggestions example
- Changed Client definition which was causing an error

000panther commented:
Hi! I tried the typescript definitions, but get the following error when importing using ES6 syntax like this

`import { AppSearchClient } from '@elastic/app-search-node';`

TS2617: 'AppSearchClient' can only be imported by using 'import AppSearchClient = require("@elastic/app-search-node")' or by turning on the 'esModuleInterop' flag and using a default import.

I also tried default imports, but I do not have the esModuleInterop flag set.

This works:
`import AppSearchClient = require('@elastic/app-search-node');`

Reading the docs it seems like this is a valid way to do things, but it's pretty uncommon and should maybe be mentioned in the docs. Maybe a Typescript default export could be used? 

Just some Feedback, thanks for providing Type Support!


```
    return new TSError(diagnosticText, diagnosticCodes)
           ^
TSError: ⨯ Unable to compile TypeScript:
src/commands/example.ts:16:10 - error TS2617: 'AppSearchClient' can only be imported by using 'import AppSearchClient = require("@elastic/app-search-node")' or by turning on the 'esModuleInterop' flag and using a default import.

 import { AppSearchClient } from '@elastic/app-search-node';
```

JasonStoltz commented:
@000panther I'll investigate, thank you for reporting this!

JasonStoltz committed:
Merge branch 'master' into ts-definitions

JasonStoltz committed:
7.16.1-beta.0

JasonStoltz committed:
Removed files from package.json

JasonStoltz committed:
7.16.1-beta.1

JasonStoltz committed:
Add not about TS environment

JasonStoltz commented:
So I removed the `files` directive so that the `.d.ts` file is published.

I then published this in a new version: 7.16.1-beta.1.

I also made a note about importing as @000panther until we can publish this with proper es module support.

JasonStoltz committed:
semi-colons

joemcelroy reviewed (approved):
looks good!

glennvgastel commented on app-search-node.d.ts:
```diff
@@ -0,0 +1,812 @@
+      apiKeyName: string,
+      options?: Record<string, any>,
+      signOptions?: Record<string, any>
+    ): void;
```
This method returns `jwt.sign`, which returns a `string` as it is used in this lib. `void` is only returned when a callback is given.

JasonStoltz replied:
Thanks for pointing that out, I'll update it.

Diluka commented:
@JasonStoltz export the internal client for the missing api usage, for exmaple #54

JasonStoltz commented:
@Diluka I thought I had exported it, let me double check.

qqpann commented:
Any updates on this? ~In version 8.3.2~ it is still not typed. :(

glenncampioni commented:
> Any updates on this? In version 8.3.2 it is still not typed. :(

@qqpan I wouldn't get my hopes of this PR getting merged up, this package has been deprecated.
You could always fork and publish it yourself, of course.

qqpann commented:
> this package has been deprecated.

Sad but this was helpful info preventing me from wasting more time. I ended up using `@elastic/elasticsearch` for my use case. Thanks anyways!
